### PR TITLE
Incorporate changes recommended by risa2000

### PR DESCRIPTION
**Pull request by cmbruns**

### CODE REVIEW

**Comment by cmbruns** [comment]
@risa2000 Could you please review this pull request? I incorporated the ideas from your pull requests, but I modified the translator so the changes would apply more generally.

Sorry I made a lot of whitespace changes, because I wanted all line endings to be newlines. It's probably much easier for you to review if you ignore whitespace changes by adding `?w=1` to the URL, like https://github.com/cmbruns/pyopenvr/pull/66/files?w=1

File: src/openvr/__init__.py [hunk 1]
```diff
@@ -1,6947 +1,6949 @@
...
-VRApplicationError_LaunchFailed = ENUM_VALUE_TYPE(109)
-VRApplicationError_ApplicationAlreadyStarting = ENUM_VALUE_TYPE(110)
-VRApplicationError_LaunchInProgress = ENUM_VALUE_TYPE(111)
-VRApplicationError_OldApplicationQuitting = ENUM_VALUE_TYPE(112)
-VRApplicationError_TransitionAborted = ENUM_VALUE_TYPE(113)
-VRApplicationError_IsTemplate = ENUM_VALUE_TYPE(114)
-VRApplicationError_SteamVRIsExiting = ENUM_VALUE_TYPE(115)
-VRApplicationError_BufferTooSmall = ENUM_VALUE_TYPE(200)
-VRApplicationError_PropertyNotSet = ENUM_VALUE_TYPE(201)
-VRApplicationError_UnknownProperty = ENUM_VALUE_TY
```
Comment (risa2000) [inline src/openvr/__init__.py]: The original API function has the following footprint:
```c++
virtual bool GetDefaultApplicationForMimeType( const char *pchMimeType, VR_OUT_STRING() char *pchAppKeyBuffer, uint32_t unAppKeyBufferLen ) = 0;
```
I.e. it returns `bool` and there seems to be no way to figure out the right buffer size, or what the return value means, unless one tries it out. There is also no `error` value in the function declaration either. So this type of function is not correctly translated (there are more of them).

Comment (risa2000) [inline src/openvr/__init__.py]: This one is another non-conforming case:
```c++
virtual EVRApplicationError GetStartingApplication( VR_OUT_STRING() char *pchAppKeyBuffer, uint32_t unAppKeyBufferLen ) = 0;
```
which returns the error code, instead of the buff size.
I believe the translator (model) logic, needs some additional test, if either the parameters *or* the return value, is of "error" type then, issue generate support for it, otherwise, it needs some additional heuristic (hardcoded edge cases :)) to resolve the right footprint.

Comment (risa2000) [inline src/openvr/__init__.py]: Another one returning void:
```c++
virtual void GetString( const char *pchSection, const char *pchSettingsKey, VR_OUT_STRING() char *pchValue, uint32_t unValueLen, EVRSettingsError *peError = nullptr ) = 0;
```

Comment (risa2000) [inline src/openvr/__init__.py]:
```c++
virtual bool ExportLiveToBuffer( VR_OUT_STRING() char *pBuffer, uint32_t *pnBufferLength ) = 0;
```

Comment (risa2000) [inline src/openvr/__init__.py]:
```c++
virtual EVROverlayError GetOverlayTransformTrackedDeviceComponent( VROverlayHandle_t ulOverlayHandle, TrackedDeviceIndex_t *punDeviceIndex, VR_OUT_STRING() char *pchComponentName, uint32_t unComponentNameSize ) = 0;
```

Comment (risa2000) [inline src/openvr/__init__.py]:
```c++
virtual EVRInputError GetBoneName( VRActionHandle_t action, BoneIndex_t nBoneIndex, VR_OUT_STRING() char* pchBoneName, uint32_t unNameBufferSize ) = 0;
```

Comment (risa2000) [inline src/openvr/__init__.py]:
```c++
virtual EVRInputError GetOriginLocalizedName( VRInputValueHandle_t origin, VR_OUT_STRING() char *pchNameArray, uint32_t unNameArraySize, int32_t unStringSectionsToInclude ) = 0;
```

Comment (risa2000) [inline src/openvr/__init__.py]:
```c++
virtual EVRSpatialAnchorError GetSpatialAnchorDescriptor( SpatialAnchorHandle_t unHandle, VR_OUT_STRING() char *pchDescriptorOut, uint32_t *punDescriptorBufferLenInOut ) = 0;
```

File: src/tests/test_string_property.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#!/bin/env python
+
+import unittest
+
+import openvr
+
+
+class TestStringProperties(unittest.TestCase):
+    def setUp(self):
+        self.vr_sys = openvr.init(openvr.VRApplication_Other)
```
[the remark this answers is not in the document]
Comment (cmbruns) [reply src/tests/test_string_property.py]: For this particular simple test, `VRApplication_Other` seems to be working fine. So I'm inclined to keep it here. Eventually we should have multiple tiers of automated tests: some which work everywhere, including in continuous integration lacking any VR hardware. Some more advanced tests might require a particular setup like already running SteamVR/Vive or whatever.

File: src/tests/test_string_property.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+    def tearDown(self):
+        pass
```
[the remark this answers is not in the document]
Comment (cmbruns) [reply src/tests/test_string_property.py]: agreed

File: src/tests/test_string_property.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    def get_string_raises(self):
+        #  1009 : Prop_ConnectedWirelessDongle_String = [error: TrackedProp_UnknownProperty]
+        prop = self.vr_sys.getStringTrackedDeviceProperty(0, openvr.Prop_ConnectedWirelessDongle_String)
```
[the remark this answers is not in the document]
Comment (cmbruns) [reply src/tests/test_string_property.py]: The way the automated tests are used at the moment is that I run them manually a few times on a functioning SteamVR/Vive system while preparing a release. If in the future this test fails in some other context, we'll fix the test at that time.

File: src/translate/generator.py [hunk 5]
```diff
@@ -1,367 +1,378 @@
...
+                pass
+            
+            
+            class ErrorCode(OpenVRError):
+                error_index = dict()
+                is_error = True  # FooError_None classes should override this
+            
+                @classmethod
+                def check_error_value(cls, error_value, message=''):
+                    error_class = cls.error_index[int(error_value)]
```
[the remark this answers is not in the document]
Comment (cmbruns) [reply src/translate/generator.py]: I'm not sure. Maybe if the `error_value` were still a `ctypes.c_int16` or whatever, it might or might not function correctly as an array index. If it works correctly without the int constructor we can remove it. Otherwise in general I'm inclined to leave mysterious things like that in place, on the assumption that at one time I understood that it was needed.

File: src/openvr/__init__.py (hunk 1 above)
Comment (cmbruns) [reply src/openvr/__init__.py]: We should spin these cases out into a separate issue. Hopefully we can find or construct some examples of these functions being used correctly from C++, so we can reason properly about how they should be used from python. It could also be helpful to see what the C# bindings do for these functions. 

Most of these functions have the `VR_OUT_STRING` clang annotation, which is great, if it is used consistently, because the translation module uses clanglib and these annotations are available at our fingertips.

File: src/tests/test_string_property.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (risa2000) [reply src/tests/test_string_property.py]: Ok

File: src/tests/test_string_property.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (risa2000) [reply src/tests/test_string_property.py]: Ok

File: src/translate/generator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (risa2000) [reply src/translate/generator.py]: I was probably interested in this "limit" case. On my install, I did not need to cast the value, so I thought maybe on some others (or in different Python versions), one had to. Anyway, it was just a curiosity.

File: src/openvr/__init__.py (hunk 1 above)
Comment (risa2000) [reply src/openvr/__init__.py]: To be honest, I have no idea, how these functions are supposed to be used correctly. The header itself does not give much of the clue and I do not have any experience with them either. The only thing I can currently think of is asking at OpenVR project, or do some wild guessing based on anecdotal experience. How would like to handle it?
